eclass/secureboot.eclass

```diff
@@ -0,0 +1,171 @@
+		ewarn "SECUREBOOT_SIGN_CERT not set. Not signing efi files."
+		return 1
+	else
+		return 0
+	fi
+}
+
+# @FUNCTION: secureboot_pkg_setup
+# @DESCRIPTION:
+# Checks if required user variables are set before starting the build
+secureboot_pkg_setup() {
+	use secureboot || return
+	debug-print-function ${FUNCNAME[0]} "${@}"
+	if [[ ${MERGE_TYPE} != binary ]]; then
+		_secureboot_warn_if_unset
+	fi
+}
+
+# @FUNCTION: secureboot_sign_efi_file
+# @USAGE: <input file> <output file>
+# @DESCRIPTION:
+# Sign a file using sbsign and the requested key/certificate.
+# If the file is already signed with our key then skip.
+secureboot_sign_efi_file() {
+	use secureboot || return
+	debug-print-function ${FUNCNAME[0]} "${@}"
+	local input_file=${1}
+	local output_file=${2}
+
+	_secureboot_warn_if_unset || return
+
+	if has_version app-crypt/sbsigntools; then
+		ebegin "Signing ${input_file}"
+		local return=1
+		if sbverify "${input_file}" --cert ${SECUREBOOT_SIGN_CERT} &> /dev/null; then
+			ewarn "${input_file} already signed, skipping"
+			return
+		else
+			local args=" --key=${SECUREBOOT_SIGN_KEY} --cert=${SECUREBOOT_SIGN_CERT} "
+			if [[ ${SECUREBOOT_SIGN_KEY} == pkcs11:* ]]; then
+				args+=" --engine=pkcs11 "
+			else
+				[[ -f ${SECUREBOOT_SIGN_KEY} ]] || die \
+					"SECUREBOOT_SIGN_KEY=${SECUREBOOT_SIGN_KEY} not found"
+			fi
+			[[ -f ${SECUREBOOT_SIGN_CERT} ]] || die \
+					"SECUREBOOT_SIGN_CERT=${SECUREBOOT_SIGN_CERT} not found"
+
+			sbsign ${args} ${input_file} --output ${output_file}
+			return=${?}
+		fi
+		eend ${return} || die -n "Signing ${input_file} failed"
+	else
+		die "secureboot signing requested but app-crypt/sbsigntools not installed"
+	fi
+}
+
+# @FUNCTION: secureboot_auto_sign
+# @USAGE: [--in-place | --ext <file extension>]
+# @DESCRIPTION:
+# Automatically discover and sign efi files in the image directory.
+#
+# By default signed files gain the .signed extension. If the --in-place
+# argument is given the efi files are replaced with a signed version in place.
+# The --ext argument allows for overriding the default .signed file extension.
+secureboot_auto_sign() {
+	use secureboot || return
+	debug-print-function ${FUNCNAME[0]} "${@}"
+
+	[[ ${EBUILD_PHASE} == install ]] ||
+		die "${FUNCNAME[0]} can only be called in the src_install phase"
+
+	local path="${ED}"
+	local -a efi_execs
+	[[ -d ${path} ]] && mapfile -td '' efi_execs < <(
+		find "${path}" -type f -name '*.efi' -print0 || die
+	)
+	(( ${#efi_execs[@]} )) ||
+		die "${FUNCNAME[0]} was called but no efi executables were found"
+
+	local ext
+	if [[ "${1}" == "--in-place" ]]; then
+		ext=""
+	elif [[ "${1}" == "--ext" ]]; then
+		[[ -z "${2}" ]] || die "Expected second argument with option --ext"
+		ext="${2}"
+	elif [[ -n "${1}" ]]; then
+		die "Invalid argument ${1}"
+	else
+		ext=".signed"
+	fi
+
+	for efi_exec in "${efi_execs[@]}"; do
+		secureboot_sign_efi_file "${efi_exec}" "${efi_exec}${ext}"
+	done
+}
+
+EXPORT_FUNCTIONS pkg_setup
+
+_SECUREBOOT_ECLASS=1
```

Comment:
Most eclasses set this right at the start if the `if`  block rather than bottom, wouldn't hurt to be consistent with these.